### ISSUE TITLE
port for fabric?

### ISSUE THREAD

**Issue by Diclo** [Pokechu22/WorldDownloader#133]
Good, this was just to mention the idea of ​​why the WDL port was not made to fabric.

**Comment by Pokechu22** [comment]
I'm probably not going to target fabric due to the differences in class names.  I'm more likely to target forge for 1.14, though I don't have a specific time when that will happen.

**Comment by uyjulian** [comment]
It seems like Fabric uses a different set of mappings called Yarn while Liteloader and Forge use the mappings provided by MCP.

**Comment by Dufftwt** [comment]
Yea true, I'd really like a version for fabric, that would be sooo useful, even if it seems to be a lot of work, It could be worth it to add fabric and forge someday 🤷‍♂

**Comment by Mhowser** [comment]
+1 for a Fabric version of WDL. Yarn is completely open in terms of licencing BTW. 
https://github.com/FabricMC/yarn

**Comment by Pokechu22** [comment]
Yarn also completely mismatches all of the existing code in the mod which is why it's not likely for me to target it any time soon (forge at least will let me re-use most stuff, including the unit tests).

**Comment by gamax92** [comment]
Took a stab at making a fabric "port", copied over the old mixins from 1.12.2, updated them for 1.14.4 and yarn mappings. Mostly just a proof of concept at this point.
A possible way to make it easier to support fabric would be to just keep producing a baseedit, remap the baseedit to fabric's intermediary/named mappings, and then just compile the mixins part under fabric's loom gradle system, merge the the two together.

Here's a copy of my ported mixins:
https://github.com/gamax92/WorldDownloader-Fabric

And compiled version:
[mod-fabricwdl-4.0.5.3-mc1.14.4.jar](https://github.com/gamax92/WorldDownloader-Fabric/releases/download/4.0.5.3/mod-fabricwdl-4.0.5.3-mc1.14.4.jar)

**Comment by Diclo** [comment]
work likes

**Comment by Mhowser** [comment]
Thank you @gamax92 This is a HUGE step forward!

**Comment by Pokechu22** [comment]
Huh, I hadn't thought about doing it like that; that seems like a pretty convenient approach.  Great idea!

**Comment by NickAcPT** [comment]
That solution seems like a good approach. Would love to see it being used officially though.

**Comment by Pokechu22** [comment]
I'll look into integrating something like this into the main code over the weekend.  The main challenge is automating the process, but I think that's doable (I already have a fair bit of weird remapping code in begradle; a bit more won't hurt).  The main difficulty with this approach is that you don't have source code you can edit for fabric; you can only edit the base edit code.  But I don't think that matters too much (and the only other approach I could thing (remapping the source code) would still have generated files you wouldn't be able to edit).  There are also the unit tests, but I think it would be possible to just compile them and remap them like the main mod code.

In the mean time: @gamax92, would you mind if I linked that in the main forum post (or maybe just a reply to it, not sure which is better for this) as an experimental build?

**Comment by Mhowser** [comment]
Currently, the WDL button doesn't align with the rest of the menu UI, I believe this is caused by [ModMenu](https://www.curseforge.com/minecraft/mc-mods/modmenu).
![Screenshot from 2019-10-11 02-29-49](https://user-images.githubusercontent.com/24195572/66669775-10acc980-ec0d-11e9-8129-765fafd1e830.png)

Without ModMenu: 
![image](https://user-images.githubusercontent.com/24195572/66670076-bbbd8300-ec0d-11e9-8899-6b7a4d2dba95.png)

**Comment by uyjulian** [comment]
Hmm, I never thought about renaming MCP mappings to Fabric mappings in the jar files. I was thinking about using the MCP mappings directly in Fabric, so there isn't an additional renaming step.

**Comment by gamax92** [comment]
The jar I made will complain about invalid hashes since well, they're all been converted to intermediary mappings, but feel free to link it. Using the MCP mappings in Fabric still has the issue that the baseedit and the old liteloader mod used minecraft jar naming while fabric mods end up as intermediary namings.

Also important to note, I have to manually edit the refmap atm since I can't get the mixin to target the GameMenuScreen.method_19836()

EDIT: I realized this was probably due to my wdl jar having minecraft classes inside of it, so this will likely not be an issue for you once it's properly integrated into the build system

**Comment by NickAcPT** [comment]
Sorry to annoy, but has there been any more progress on this issue?

**Comment by Pokechu22** [comment]
I've been pretty busy lately and haven't made much progress, but I am working on it as I have time.

**Comment by Mhowser** [comment]
Thanks for the reply, we do appreciate your effort!

**Comment by Mhowser** [comment]
Hi again, has there been any progress since last time?

**Comment by Pokechu22** [comment]
Unfortunately not much; I've been busy with classes again.  I'll try to get some progress made over the coming long weekend.

**Comment by Mhowser** [comment]
That's okay, real life comes first ;), thanks for the update!

**Comment by Xanton19** [comment]
I don't want to bother you because you are busy and this question has been asked several times in this thread, but will there be any potential fabric 1.15.2 version of WDL? As the code that @gamax92 has in his Fabric WDL Repo is for Minecraft 1.14.4. Anyways, thanks for the efforts in developing WDL.

**Comment by Pokechu22** [comment]
I plan on targeting both 1.14.4 and 1.15.2 with the port; the same technique should work for both.  I don't think it'd be super necessary to target the other minor versions in the ports since I don't think forge or fabric are maintaining them (and I don't want to introduce 16 new versions that all have to be built when most wouldn't be useful; the base edit build should cover the rare case that a server is on one of those minor versions).

**Comment by Xanton19** [comment]
Thank you for your answer and for sharing the plans on fabric.

If I correctly understand about Minor versions (For example 1.14-1.14.3 and 1.15-1.15.1), I totally agree with you that they are not needed for Fabric/Forge Builds.
Are you referring to @gamax92 technique on porting the mod to fabric?

Thank you for your time and have a great weekend.

**Comment by Boobies** [comment]
Any news on this? @gamax92 does not seem to be updating his branch ever since the initial commit.

**Comment by Pokechu22** [comment]
Getting 1.16.1/1.16.2 working is the biggest priority, which will be followed by getting 1.16.x forge working (since that shouldn't take much more effort once 1.16.x base-edit works), and then I'll work on this.  I've been busy lately and don't have an ETA for any of these steps though.

**Comment by IMS212** [comment]
Any news? I have been trying to port @gamax92 's version to 1.16.3 now, but i'm not very experienced and I don't know how to update mixins. _I believe_ I have the rest of the stuff ported, but I cannot test it.

**Comment by Pokechu22** [comment]
I've been busy with class lately so I haven't made many changes.  I do hope to get a proper fabric build out along the same time 1.16.4 releases but I can't be sure of it (a current bigger goal is to get the protocol documentation on wiki.vg merged and up to date since the current situation there is pretty jank).

**Comment by yass03** [comment]
Hi! it's been some time since your last message,anything new? :)

**Comment by uyjulian** [comment]
If there isn't any new message, there is no new news.

**Comment by Pokechu22** [comment]
Yeah, unfortunately no new news with regards to fabric.  I've done some work on the base 1.16.4 version, and wiki.vg is *mostly* updated now (for both releases and snapshots, though I still need to do a check-over to make sure things are merged), but I'm still quite busy and haven't made much progress here.  I know this is a thing that a lot of people want, and I wish I were less busy, but I don't know when that'll happen.

**Comment by Mhowser** [comment]
Do you know any developers in your circle that could assist in getting this
finished?

**Comment by Mhowser** [comment]
@gamax92 Is there any possibility that you could update the mixins you ported over to 1.16.4?

**Comment by NiyahVE** [comment]
That would be amazing ! Keep the good work, changing from fabric to forge for this is alright for the moment but a tiny bummer  :)

**Comment by merrx** [comment]
I just found #206 and this issue seems to be the current fabric WDL thread, so I'd like to leave my experience here.
It is possible to use the base-edit WDL (via MultiMC) with serveral other fabric mods and most of them work fine.
I only found two mods, which don't work:
* Multiconnect (crashes at directly Startup)
* Sodium (here the downloaded world is always of void-type with no saved chunks)

I just wanted to leave a little context here. Sodium, Lithium and Phosphor are performance mods using the fabric modloader. In this case sodium is the curpit, as it changes how chunks are processed. Temporaily disableing this Sodium allows base-edit WDL to function properly.

**Comment by IMS212** [comment]
also optifabric causes issues, hitting download world crashes game

**Comment by AngelFireLA** [comment]
good luck

**Comment by joeyjumper94** [comment]
there is going to be huge demand for a fabric version soon

**Comment by Mhowser** [comment]
There pretty much already is a high demand, the problem is that the dev(s) do not have the time to work on this unfortunately.

**Comment by PK268** [comment]
Hey sorry for bugging you 2 years after OP but is there any new updates? I'd be really cool if this worded out.

**Comment by Pokechu22** [comment]
2 years... yikes, time has gone by quickly.  I've still been fairly busy lately and still am backlogged with several things (including updating the 1.17 protocol documentation, which is now 11 weeks out of date), but this is still something that I want to work on.  Apologies to everyone who's waiting on this that I still don't know when I'll have time to work on it.

**Comment by TheAwesome98-Real** [comment]
> 2 years... yikes, time has gone by quickly. I've still been fairly busy lately and still am backlogged with several things (including updating the 1.17 protocol documentation, which is now 11 weeks out of date), but this is still something that I want to work on. Apologies to everyone who's waiting on this that I still don't know when I'll have time to work on it.

not having time to work on stuff is fine lol

**Comment by IMS212** [comment]
I've made a full Fabric 1.16.5 port here: https://github.com/IMS212/WorldDownloader/tree/fabric

Compiled build: https://github.com/IMS212/WorldDownloader/raw/fabric/mod-fabricwdl-4.1.1.0-mc1.16.5.jar

**Comment by Pokechu22** [comment]
Well done!  What technique did you use for it?  Manual porting of the code to use a new naming scheme, or something more fancy?

**Comment by IMS212** [comment]
Manual porting of the code (still using MCP mappings), then an automatic rename to mojang mappings. This is definitely one of the harder ways to do it, but it seems to work quite well.

**Comment by IMS212** [comment]
Sadly, i've managed to find many issues with the port i just made, so i don't recomend using it full time for now. (It can download worlds fine, but trying to play with it causes issues.)

**Comment by Mhowser** [comment]
Thank you do much for doing this @IMS212 any chance you can compile it for 1.17?

**Comment by IMS212** [comment]
Sadly, 1.17 seems to have many other issues I can’t fix currently.

**Comment by jiaaluo** [comment]
Nice thanks for the port! I myself have been using a heavily modified version of bobby that allows auto-downloading of all worlds.
Unfortunately with these modifications, only blocks are stored, no inventories or other data though sodium is supported.

**Comment by FlorianMichael** [comment]
[https://uploadi.ng/enzaxd/iSOdo93V](url) I made a Fabric 1.17.1 port for the Yarn Mappings

**Comment by IMS212** [comment]
@FlorianMichael nice, would you mind publishing it?

**Comment by FlorianMichael** [comment]
No, it's built into my private utility client and I'm not going to release it, but it's not that much work to port it either.

**Comment by Mhowser** [comment]
What was the point of telling us you made a 1.17 port if you aren't going to share that code?

**Comment by NickAcPT** [comment]
There's this on the WDL license..
> The Author must make the Mod source freely and publicly available.

**Comment by Pokechu22** [comment]
The license only applies when distributing a binary (in other words, if you make something for yourself, but never share it, you don't need to share the source code to it - most licenses work that way).  But also, if the porting process went quickly, it probably doesn't handle all of the edge cases that make a 1.17 port (or a fabric port) difficult, such as the adjusted world height.

I do agree that there isn't much of a point in telling us that you made a port if you're not doing anything else with it, though.

**Comment by IMS212** [comment]
^ at this point you’re just doing this to piss us off, showing that you have it and are using it to cheat

**Comment by triphora** [comment]
EULA noncompliant and license noncompliant. Nice.

**Comment by NickAcPT** [comment]
> The license only applies when distributing a binary (in other words, if you make something for yourself, but never share it, you don't need to share the source code to it - most licenses work that way).

Well, it really depends, but from what I've seen, licenses like these apply to the code itself, not the binary. I've had a read of the current license and I find it a bit ambiguous and I find something like GPL a bit be less ambiguous regarding this.

**Comment by IMS212** [comment]
and now I also noticed you’re using iris, which was made for the purpose of open source shader mods, where we encourage open source, in your closed hacked client. Just... why?

**Comment by fuck-shithub** [comment]
It's typical western game cheating culture. Private cheats, no source code.

**Comment by OakBricks** [comment]
fucking idiotic, if its private can you at least not tell anyone, nobody on this project wants to download your fucking hacked client that not only breaks the license of WDL, but also breaks the EULA, and with the Iris issue, in the LGPLv3 license there is a section on combined works, im going to assume that your hacked client is a combined work (it probably is because you got both iris and WDL in a what appears to be bodged together hacked client) and how you need to give a notice on what projects are used in your project, which im assuming you don't.

just please contribute if you are showcasing your WDL fork in your hacked client.

**Comment by Mhowser** [comment]
> just please contribute if you are showcasing your WDL fork in your hacked client.

It's the least they can do after getting all our hopes up...

**Comment by NatoBoram** [comment]
@Pokechu22 Can the recent comments about @FlorianMichael be marked as off-topic?

**Comment by Pokechu22** [comment]
Hidden.

And, again, sorry for how long everything is taking me.  I've been busy with a lot of projects lately.

**Comment by Mhowser** [comment]
Would a Bug Bounty  be a good incentive for this issue?

**Comment by Pokechu22** [comment]
I don't think it would be, sorry.

**Comment by Texbio** [comment]
:/

**Comment by Xplodin** [comment]
> No, it's built into my private utility client and I'm not going to release it, but it's not that much work to port it either.

nice job, you've officially earned cunt status

**Comment by florianreuth** [comment]
> > No, it's built into my private utility client and I'm not going to release it, but it's not that much work to port it either.
> 
> nice job, you've officially earned cunt status

Why are you replying to a message that is already over 2 years old?

**Comment by Xplodin** [comment]
> > > No, it's built into my private utility client and I'm not going to release it, but it's not that much work to port it either.
> > 
> > 
> > nice job, you've officially earned cunt status
> 
> Why are you replying to a message that is already over 2 years old?

Because its still relevant today.
Im already porting the mod for fabric 1.19.2 anyways as its not much but im gladly releasing it for public use

**Comment by TheAwesome98-Real** [comment]
> > > > No, it's built into my private utility client and I'm not going to release it, but it's not that much work to port it either.
> > > 
> > > 
> > > nice job, you've officially earned cunt status
> > 
> > 
> > Why are you replying to a message that is already over 2 years old?
> 
> Because its still relevant today. Im already porting the mod for fabric 1.19.2 anyways as its not much but im gladly releasing it for public use

and the source code? :3

**Comment by florianreuth** [comment]
> > > > No, it's built into my private utility client and I'm not going to release it, but it's not that much work to port it either.
> > > 
> > > 
> > > nice job, you've officially earned cunt status
> > 
> > 
> > Why are you replying to a message that is already over 2 years old?
> 
> Because its still relevant today. Im already porting the mod for fabric 1.19.2 anyways as its not much but im gladly releasing it for public use

no, my stupid post is not relevant today anymore

**Comment by ThatDevExotic** [comment]
> > > > > No, it's built into my private utility client and I'm not going to release it, but it's not that much work to port it either.
> > > > 
> > > > 
> > > > nice job, you've officially earned cunt status
> > > 
> > > 
> > > Why are you replying to a message that is already over 2 years old?
> > 
> > 
> > Because its still relevant today. Im already porting the mod for fabric 1.19.2 anyways as its not much but im gladly releasing it for public use
> 
> no, my stupid post is not relevant today anymore

lol
